keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,579 @@
+
+```
+// Controller code
+
+func (cjc *CronJobController)  Run() {
+	...
+    for i := 0; i < workers; i++ {
+        go wait.Until(cjc.worker, time.Second, stopCh)
+    }
+...
+}
+
+```
+
+##### Handling Cron aspect
+To detect which CronJob has met its schedule and need to create Jobs we need to implement a timer component. These are the possible options for implementing the timer:
+
+| algorithm  | how it works | notes |
+|:----------|:----------|:-------|
+| Unordered timer list | Periodic Sweep from cache | Slower and similar to existing implementation. But improved because we sweep fom the cache instead of API server | 
+|Ordered timer list| Maintain ordered list of Cronjob keys and next time of expiry. Keep starting a timer with the earliest expiry. | Efficient. Reinsertion to list takes O(n) |
+|Timer trees| Instead of ordered list use a sorted binary tree. | More efficient. Insertion is O(log n) |
+|Heap based timer|A variant of ordered timer list where heap is used to store the next expiry time | Efficient compared to ordered list. Bookkeeping and insertion is O(log n). |
+|Simple Timing wheels| circular buffer of MaxTimeOut slots. List of expiring timers at each slot. | Works for small bounded  MaxTimeOut which is not our case. Insertion and removal is O(1) via indexing |
+|Hashed Wheel| Hash expiring time and insert in a hash table with linked list at each index | Bookkeeping is O(1) and worst case insertion is O(n) |
+|Hierarchical Wheel| multiple timer wheels for different resolutions (Seconds, minutes, hours, days). When seconds rolls over we grab the next minutes timers and recreate the seconds wheel. similarly for minutes and hours. | Sharding at different hierarchy levels improves insertion and bookkeeping performance. |
+
+
+We shall implement a Heap based timer algorithm. We will introduce a separate queue with the [`DelayingInterface`](https://github.com/kubernetes/client-go/blob/master/util/workqueue/delaying_queue.go#L37) that implements heap based single shot api [`AddAfter`](https://github.com/kubernetes/client-go/blob/master/util/workqueue/delaying_queue.go#L150). Every time we process an entry from this queue, we will add it back to the queue to simulate a periodic timer.
+
+
+For further reading:
+1. [Reinventing timer wheel](https://lwn.net/Articles/646950/)
+2. [Hashed and hierarchical timer wheel](http://www.cs.columbia.edu/~nahum/w6998/papers/sosp87-timing-wheels.pdf)
+3. [Golang timers in multi-cpu systems](https://github.com/golang/go/commit/76f4fd8a5251b4f63ea14a3c1e2fe2e78eb74f81)
+4. [Go timerwheel](https://github.com/RussellLuo/timingwheel)
+
+##### Metrics
+We propose to add metrics that could expose the performance and health of the controller including and not limited to: 
+- Skew (actualJobCreationTime-expectedJobCreationTime)
+- Queue depth (pending CronJob and Job entries in the Queue)
```

Comment:
IIRC, you get this for free from the framework, right? If so, make that explicit.

Reply:
yes. made it explicit.